### ISSUE TITLE
Question: error on ligand is outside the grid box

### ISSUE THREAD

**Issue by mosizei** [ccsb-scripps/AutoDock-Vina#112]
I got this error massage when I used the python vina==1.2.3:

```
ERROR: The ligand is outside the grid box. Increase the size of the grid box or center it accordingly around the ligand.
```
How can  I solve this error? 
There were no error while I was working with the previous version Vina 1.2.2 (not python). Although the ligands wasn't inside of the grid box Vina was able to dock it to the receptor without any problem.

**Comment by diogomart** [comment]
You probably want to use `dock()` instead of `score()`.

**Comment by sxl324** [comment]
I got the same error for one protein. It was fixed only by reinstall vina==1.2.2 from 1.2.3.....nothing changed in my code. I do not know what was wrong.

**Comment by mosizei** [comment]
I have done the same, downgrading to vina==1.2.2 was the solution. :)

**Comment by diogomart** [comment]
There is a large penalty for ligands that are outside the box. The goal of that penalty is to bring the ligands back into the box during docking.

v1.2.3 checks if the ligand is inside the box before scoring it.

v1.2.2 scores the ligand even if it is outside the box. The numbers it returns are useless for any purpose except informing about the distance between the ligand and the box.

Here's the output of `v1.2.3` while scoring a ligand inside and outside the box:
```
vina version: 1.2.3
Computing Vina grid ... done.
ligand center:   -0.011   -0.218    0.005
score: [29.066 29.066  0.     0.     0.     0.     0.     0.   ]
ligand center:   99.989   99.782  100.005
ERROR: The ligand is outside the grid box. Increase the size of the grid box or center it accordingly around the ligand.
```

and the same with `v1.2.2`
```
vina.__version__ not available:
Computing Vina grid ... done.
ligand center:   -0.011   -0.218    0.005
score: [29.066 29.066  0.     0.     0.     0.     0.     0.   ]
ligand center:  199.989  199.782  200.005
score: [3.507528e+09 3.507528e+09 0.000000e+00 0.000000e+00 0.000000e+00
 0.000000e+00 0.000000e+00 0.000000e+00]
```

and the script
```python
from rdkit import Chem
from rdkit.Chem import rdDistGeom
from meeko import MoleculePreparation
from vina import Vina
import numpy as np
import vina

try:
    print("vina version:", vina.__version__)
except AttributeError:
    print("vina.__version__ not available:")

receptor_string = "ATOM      1  C   UNL     1       0.000   0.000   0.000  0.00  0.00    +0.000 C "
rec_fn = "my_rec_is_just_a_carbon.pdbqt"
print(receptor_string, file=open(rec_fn, "w"))

maps_center = [0, 0, 0]
maps_size_angstroms = [10, 10, 10]

pyridine = Chem.MolFromSmiles("C1=CC=CN=C1")
pyridine = Chem.AddHs(pyridine)
etkdg_params = rdDistGeom.ETKDGv3()
rdDistGeom.EmbedMolecule(pyridine, etkdg_params)

mkprep = MoleculePreparation()
mkprep.prepare(pyridine)
lig_string = mkprep.write_pdbqt_string()

v = Vina()
v.set_receptor(rec_fn)
v.set_ligand_from_string(lig_string)
v.compute_vina_maps(center=maps_center, box_size=maps_size_angstroms)

def print_ligand_center(molsetup):
    lig_xyz = []
    for atom_index, is_atom_ignored in molsetup.atom_ignore.items():
        if not is_atom_ignored:
            lig_xyz.append(molsetup.coord[atom_index].copy())
    lig_xyz = np.array(lig_xyz)
    print("ligand center: %8.3f %8.3f %8.3f" % tuple(np.mean(lig_xyz, 0)))

print_ligand_center(mkprep.setup)
score = v.score()
print("score:", score)

# translate ligand out of box
for atom_index in mkprep.setup.coord:
    x, y, z = mkprep.setup.coord[atom_index]
    mkprep.setup.coord[atom_index] = np.array([x + 200, y + 200, z + 200])

print_ligand_center(mkprep.setup)
lig_string = mkprep.write_pdbqt_string()
v.set_ligand_from_string(lig_string)
score = v.score()
print("score:", score)
```

**Comment by biao-ma** [comment]
@diogomart 
Thanks for the explanation, I understand why the error occurs, but I still don't know how to avoid this error. You said "the purpose of this penalty is to bring the ligand back into the box during the docking process", but how do you bring the ligand back into the box during the docking process? Is it necessary to add a judgment to the script and force the ligand back into the box when an error occurs?

**Comment by diogomart** [comment]
To avoid the error when scoring with `score()` or `--score_only`, specify a box size and center that fully encompasses the ligand.

`v.dock()` will automatically bring ligands into the box, but not `v.score()` as it does not move the ligand.

**Comment by JonasLi-19** [comment]
Thanks for explaining the v.score and its thoughtful design, but when I use this to score the pocket.pdb and ligand.mol2 from PDBbind dataset, it would still raises this error. However, the molecule and protein file in PDBbind dataset are dealt to be correctly combined, and I use the center of ligand to be the map center. why the ligand is still out of the box?

**Comment by diogomart** [comment]
I don't know. Can you post the files to reproduce?

**Comment by jaehongshin** [comment]
Hi, I got the same errors. I tried to add an error handling but it did not work.  To make high-throughput docking with python binding, it would be nice to have an option to avoid exiting the program, when the vina turns this error.

**Comment by daoquangtung2411** [comment]
Hi, has anyone solved this issue. I'm has downgraded to vina 1.2.2 and it runs smoothly, but one of my colleagues cannot install vina 1.2.2 due to python version (3.11). 
If anyone has solved this, please share with us.
Thank in advanced

**Comment by daoquangtung2411** [comment]
> To avoid the error when scoring with `score()` or `--score_only`, specify a box size and center that fully encompasses the ligand.
> 
> `v.dock()` will automatically bring ligands into the box, but not `v.score()` as it does not move the ligand.

Hi, I encounter the error even when increase size to 126x126x126 and spacing = 1.000. Can you further explain the solution

**Comment by diogomart** [comment]
The `score()` function does only a single point energy evaluation using the input coordinates. To dock the ligand you need to call the `dock()` function. The error was added to help users identify misplaced ligands or misplaced search spaces. It is possible that your ligand is very far from the receptor, in which case calling `score()` is meaningless. 

Why do you want to call `score()` and not `dock()` ?

**Comment by daoquangtung2411** [comment]
> The `score()` function does only a single point energy evaluation using the input coordinates. To dock the ligand you need to call the `dock()` function. The error was added to help users identify misplaced ligands or misplaced search spaces. It is possible that your ligand is very far from the receptor, in which case calling `score()` is meaningless.
> 
> Why do you want to call `score()` and not `dock()` ?

No, I call dock() just like the script in docs. But with vina 1.2.3, I have to increase box size to 200x200x200 to bring ligand inside the box. 
But the result was unsatisfied in both pose and affinity score (it bind to multiple pocket, not have the similar pose as crystallized, the affinity score was scatter broadly).

**Comment by daoquangtung2411** [comment]
> The error was added to help users identify misplaced ligands or misplaced search spaces. It is possible that your ligand is very far from the receptor, in which case calling `score()` is meaningless.

Can you explain further about position of ligand to the receptor. Because if I dock with vina 1.2.2, I received both the right pose (similar to crystallized) and satisfied affinity score. 
Thank you very much for your reply

**Comment by diogomart** [comment]
> No, I call dock() just like the script in docs. But with vina 1.2.3, I have to increase box size to 200x200x200 to bring ligand inside the box.

This is very surprising. Can you share the files and commands/scripts to reproduce?

**Comment by daoquangtung2411** [comment]
[docking_code.txt](https://github.com/ccsb-scripps/AutoDock-Vina/files/12364909/docking_code.txt)
[7myi_config.txt](https://github.com/ccsb-scripps/AutoDock-Vina/files/12364913/7myi_config.txt)
<img width="426" alt="Vina1 2 3_pose" src="https://github.com/ccsb-scripps/AutoDock-Vina/assets/78310750/24b772b1-9790-43d6-adc9-3d9f3fada565">
<img width="483" alt="Vina1 2 2" src="https://github.com/ccsb-scripps/AutoDock-Vina/assets/78310750/1a1fd16d-5067-4249-b3ad-d8148abea72d">
<img width="777" alt="Vina1 2 3_l+p" src="https://github.com/ccsb-scripps/AutoDock-Vina/assets/78310750/76966af1-076f-460b-b65f-759e2f40a654">

> > No, I call dock() just like the script in docs. But with vina 1.2.3, I have to increase box size to 200x200x200 to bring ligand inside the box.
> 
> This is very surprising. Can you share the files and commands/scripts to reproduce?

Here's the docking code and config file I use. Also picture of docking pose. I am docking BACE1 from SMILES

With vina1.2.2, after redocking from SMILES, the result pose is very similar to crystallized with protein complex. The scoree was satisfy, and rmsd of other pose is around 3.0 compard to the best pose. And I also check interaction, it the same between results and complex.

But as you can see, with vina 1.2.3, I must increase box size to 200x200x200, and ligand bind with multiple pocket. The best pose still bind to primary pocket, but its pose differ from crystallized.

**Comment by diogomart** [comment]
In your script you are calling `score()` before calling `dock()`. At that point, its position and orientation are independent from the protein, so it's not surprising it is outside the search space. Also, there's no need to call `score()` at all, as you'll have access to the energies after `dock()`. 

Just a comment: there's no need to re-instantiate the `Vina()` object and re-calculate the maps inside the `for` loop that iterates over ligands, that needs to be done only once.

**Comment by daoquangtung2411** [comment]
> In your script you are calling `score()` before calling `dock()`. At that point, its position and orientation are independent from the protein, so it's not surprising it is outside the search space. Also, there's no need to call `score()` at all, as you'll have access to the energies after `dock()`.
> 
> Just a comment: there's no need to re-instantiate the `Vina()` object and re-calculate the maps inside the `for` loop that iterates over ligands, that needs to be done only once.

Thanks a lot. I delete score() and it work perfectly.

**Comment by SaisaiSun** [comment]
Can anyone help explaining the results from running vina without the config file (with --score_only) just for re-scoring the crystallized structure from PDB?

**Comment by diogomart** [comment]
What do you mean by "without the config file"? Any option can be either in the config file or passed as a parameter.

The `--score_only` or `score()` calculate the energy of the input coordinates without any optimization.

**Comment by SaisaiSun** [comment]
> What do you mean by "without the config file"? Any option can be either in the config file or passed as a parameter.
> 
> The `--score_only` or `score()` calculate the energy of the input coordinates without any optimization.

Thanks for your reply. I mean that run vina without the center_x, center_y, center_z, size_x, size_y and size_z parameters just for re-scoring the complex. Is it workable? And are the score_only results credible by running vina without parameters mentioned above or with a random set of center_x,...,sxize_z?

**Comment by rwxayheee** [comment]
Hi @SaisaiSun, 

`I mean that run vina without the center_x, center_y, center_z, size_x, size_y and size_z parameters just for re-scoring the complex. Is it workable?`

I tried: 

`./vina --receptor receptor.pdbqt --ligand pose_1.pdbqt --score_only`

Didn't work for me. I got: 

`Scoring function : vina
Rigid receptor: receptor.pdbqt
Ligand: python_dock_1.pdbqt
Grid center: X 4.94066e-324 Y 1.03978e-312 Z 0
Grid size  : X 0 Y 0 Z 2.52517e-312
Grid space : 0.375
Exhaustiveness: 8
CPU: 0
Verbosity: 1



An error occurred: 

Vina runtime error: Grid box dimensions must be greater than 0 Angstrom`

If you do not want to assign the gridbox yourself, the following is a walkaround: 
`./vina --receptor receptor.pdbqt --ligand pose_1.pdbqt --score_only **--autobox**`

According to the output, this is how --autobox assigns the location and dimensions of grid box: 

`Grid center: ligand center (autobox)
Grid size  : ligand size + 4 A in each dimension (autobox)`

**Comment by SaisaiSun** [comment]
> Hi @SaisaiSun,
> 
> `I mean that run vina without the center_x, center_y, center_z, size_x, size_y and size_z parameters just for re-scoring the complex. Is it workable?`
> 
> I tried:
> 
> `./vina --receptor receptor.pdbqt --ligand pose_1.pdbqt --score_only`
> 
> Didn't work for me. I got:
> 
> ```
> Scoring function : vina
> Rigid receptor: receptor.pdbqt
> Ligand: python_dock_1.pdbqt
> Grid center: X 4.94066e-324 Y 1.03978e-312 Z 0
> Grid size  : X 0 Y 0 Z 2.52517e-312
> Grid space : 0.375
> Exhaustiveness: 8
> CPU: 0
> Verbosity: 1
> 
> 
> 
> An error occurred: 
> 
> Vina runtime error: Grid box dimensions must be greater than 0 Angstrom
> ```
> 
> If you do not want to assign the gridbox yourself, the following is a walkaround: `./vina --receptor receptor.pdbqt --ligand pose_1.pdbqt --score_only --autobox`
> 
> According to the output, this is how `--autobox` assigns the location and dimensions of grid box:
> 
> `Grid center: ligand center (autobox) Grid size : ligand size + 4 A in each dimension (autobox)`

Thank you so much. It's very helpful!

**Comment by rwxayheee** [comment]
Closing as resolved
